Library/Homebrew/cmd/autoremove.rb

```diff
@@ -0,0 +1,53 @@
+# typed: false
+# frozen_string_literal: true
+
+require "formula"
+require "cli/parser"
+require "uninstall"
+
+module Homebrew
+  extend Uninstall
+
+  module_function
+
+  def autoremove_args
+    Homebrew::CLI::Parser.new do
+      usage_banner <<~EOS
+        `autoremove` [<options>]
+
+        Remove packages that weren't installed on request and are no longer needed.
+      EOS
+      switch "-n", "--dry-run",
+             description: "Just print what would be removed."
+      named 0
+    end
+  end
+
+  def get_removable_formulae(formulae)
+    removable_formulae = Formula.installed_non_deps(formulae).reject do |f|
+      Tab.for_keg(f.any_installed_keg).installed_on_request
+    end
+
+    removable_formulae += get_removable_formulae(formulae - removable_formulae) if removable_formulae.any?
```

Comment:
Why does this need to be called recursively? Isn't it sufficient to do a single call of `installed_formulae_with_no_dependents` and then remove those that are `installed_on_request`?

Reply:
This was based on the original design. If `b` depends on `a` and nothing else depends on `a`, `a` can be safely removed after `b` is removed. If it's not recursive, it would not pick up `a`, it would only pick up `b`.

Reply:
I don't think it's necessary. If a depends on b depends on c and you check a and b and c you don't need to do any recursion to figure out that both b and c are necessary.

Reply:
In this example, `installed_formulae_with_no_dependents` will only pick up a. It won't pick up b and c because they have dependents a and b respectively. After removing a, the next call to `installed_formulae_with_no_dependents` will then pick up b.

Reply:
Ah, I see, you're wanting `auto remove` to handle them all in a single pass. Makes sense.